rclc/src/functions.c

```diff
@@ -0,0 +1,255 @@
+#include <rclc/rclc.h>
+
+#include <rcl/rcl.h>
+#include <rcl/error_handling.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/*/
+#define ALLOCATE(s) rcl_get_default_allocator().allocate(s, rcl_get_default_allocator().state)
+#define DEALLOCATE(ptr) rcl_get_default_allocator().deallocate(ptr, rcl_get_default_allocator().state)
+#define REALLOCATE(ptr, s) rcl_get_default_allocator().reallocate(ptr, s, rcl_get_default_allocator().state)
+/*/
+#define ALLOCATE(s) malloc(s)
+#define DEALLOCATE(ptr) free(ptr)
+#define REALLOCATE(ptr, s) realloc(ptr, s)
+//*/
+
+struct rclc_subscription_t
+{
+  rcl_subscription_t rcl_subscription;
+  rclc_callback_t user_callback;
+
+  rclc_node_t * node;
+};
+
+struct rclc_publisher_t
+{
+  rcl_publisher_t rcl_publisher;
+
+  rclc_node_t * node;
+};
+
+struct rclc_node_t
+{
+  rcl_node_t rcl_node;
+
+  rclc_subscription_t ** subs;
+  size_t subs_s;
+};
+
+rclc_ret_t rclc_init(int argc, char ** argv)
+{
+  rcl_ret_t ret = rcl_init(argc, argv, rcl_get_default_allocator());
+  if(ret != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_init] error in rcl_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+  return ret;
+}
+
+bool rclc_ok()
+{
+  return true;
+}
+
+void rclc_sleep_ms(size_t milliseconds)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+
+  rc = rcl_wait_set_clear_subscriptions(&wait_set);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_clear_subscriptions: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+
+  rc = rcl_wait(&wait_set, RCL_MS_TO_NS(milliseconds));
+  if (rc != RCL_RET_OK && rc != RCL_RET_TIMEOUT) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+    return;
+  }
+
+  rc = rcl_wait_set_fini(&wait_set);
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_sleep_ms] error in rcl_wait_set_fini: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+}
+
+void rclc_spin_node(rclc_node_t * node)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
```

Comment:
I think this should return an error code given that it this failed, any operation on the waitset in the while will crash.

Reply:
Yes, and should I destroy the waitset with `rcl_wait_set_fini` in this case ?

Reply:
In this case, the waitset was not created properly so you shouldn't have to destroy it.